app/routes/__init__.py

```diff
@@ -137,6 +146,48 @@ async def raster_tile_cache_version_dependency(
     return dataset, version
 
 
+async def cog_asset_dependency(
+    request: Request,
+    dataset: Optional[COGDatasets] = Query(None, description=COGDatasets.__doc__),
+    version: Optional[str] = Query(
+        None, description="Data API dataset version.", regex=VERSION_REGEX_NO_LATEST
+    ),
+    name: Optional[str] = Query(
+        "default", description="Name of the COG asset to display"
+    ),
+    url: Optional[str] = Query(
+        None,
+        description="Dataset path. This needs to be set if `dataset` and `version` query parameters for a Data API dataset are not set.",
+    ),
+) -> Optional[str]:
+
+    if dataset is None and version is None and url is None:
+        raise HTTPException(
+            status.HTTP_400_BAD_REQUEST,
+            "Need to pass either `url` or `dataset` and `version` pair for Data API dataset.",
+        )
+
+    if dataset and version and url:
+        raise HTTPException(
+            status.HTTP_400_BAD_REQUEST,
+            "Need to pass either `url` or `dataset` and `version` pair, not both.",
+        )
+
+    if dataset and version:
+        folder: str = (
+            f"s3://{DATA_LAKE_BUCKET}/{dataset}/{version}/raster/epsg-4326/cog"
+        )
+        if "bands" in request.query_params:
+            return folder
+
+        if "mosaic" in str(request.url):
+            return f"{folder}/mosaic.json"
+
+        return f"{folder}/{name}.tif"
```

Comment:
so the name param will be the implementation name for single bands?

Reply:
Yup, btw any reason we should call it implementation (which I don't quite get in the existing tile cache service context) and not just name (of the COG file)?